web/packages/teleterm/src/ui/Search/pickers/ActionPicker.tsx

```diff
@@ -281,30 +307,71 @@ const ExtraTopComponents = (props: {
   }
 };
 
+/**
+ * ActionPickerStatus helps with displaying ExtraTopComponents. It has two goals:
+ *
+ *   * Encapsulate business logic so that anything that ExtraTopComponents renders can just read
+ *     ActionPickerStatus fields.
+ *   * Represent only valid UI states. For example, inputState 'no-input' doesn't have hasNoResults
+ *     field as this field would make no sense in a situation where no search requests were made.
+ *
+ * As you may notice, ActionPickerStatus doesn't say whether the search request is in progress or
+ * not, simply because displaying the progress bar is handled by another component. The questions
+ * answered by ActionPickerStatus are valid to ask no matter what the state of the request is.
+ */
 type ActionPickerStatus =
-  | { status: 'no-input'; hasNoRemainingFilterActions: boolean }
-  | { status: 'processing' }
   | {
-      status: 'finished';
+      // no-input: The input is empty.
+      inputState: 'no-input';
+      searchMode:
+        | {
+            // no-search: The search bar is pristine, that is the input and the filters are empty.
+            kind: 'no-search';
```

Comment:
I feel kinda uneasy about having two sources of truth when it comes to `searchMode`. It is defined in `useSearch.tsx` as well:

https://github.com/gravitational/teleport/blob/ef0cb91b0397ecb71c2b36a6acddf80d5cd8acdf/web/packages/teleterm/src/ui/Search/useSearch.ts#L360-L377

Perhaps it'd make more sense for `getActionPickerStatus` to accept `searchMode` coming from `useSearch` as an argument instead of accepting both `inputValue` and `filters` and repeating the logic.

`searchMode` would need to be extracted from `useResourceSearch` though so that it can be known without waiting for the search to resolve. This could be done by exporting `getResourceSearchMode` and then passing `searchMode` to both the function returned from `useResourceSearch` and to `getActionPickerStatus`.

On top of that, those search modes are not 100% the same, `useResourceSearch` trims the input to avoid making unnecessary requests, so that logic would need to be made consistent between these two places.

Alas, this would require more time then I have at the moment. Plus `ActionPickerStatus` is still subject to change, for example when taking filters into account when displaying offline clusters (https://github.com/gravitational/teleport/pull/24190#discussion_r1162613367).